api/handler/cors.go

```diff
@@ -0,0 +1,280 @@
+	if err = xml.Unmarshal(info, cors); err != nil {
+		return
+	}
+
+	withCredentials := r.Header.Get(api.Authorization) != ""
+
+	for _, rule := range cors.CORSRules {
+		for _, o := range rule.AllowedOrigins {
+			if o == origin {
+				for _, m := range rule.AllowedMethods {
+					if m == r.Method {
+						w.Header().Set(api.AccessControlAllowOrigin, o)
+						w.Header().Set(api.AccessControlAllowCredentials, "true")
+						w.Header().Set(api.Vary, "Origin")
+						return
+					}
+				}
+			}
+			if o == wildcard && !withCredentials {
+				for _, m := range rule.AllowedMethods {
+					if m == r.Method {
+						w.Header().Set(api.AccessControlAllowOrigin, o)
+						return
+					}
+				}
+			}
+		}
+	}
+}
+
+func (h *handler) Preflight(w http.ResponseWriter, r *http.Request) {
+	reqInfo := api.GetReqInfo(r.Context())
+	bktInfo, err := h.obj.GetBucketInfo(r.Context(), reqInfo.BucketName)
+	if err != nil {
+		h.logAndSendError(w, "could not get bucket info", reqInfo, err)
+		return
+	}
+
+	origin := r.Header.Get(api.Origin)
+	if origin == "" {
+		h.logAndSendError(w, "origin request header needed", reqInfo, errors.GetAPIError(errors.ErrBadRequest))
+	}
+
+	method := r.Header.Get(api.AccessControlRequestMethod)
+	if method == "" {
+		h.logAndSendError(w, "Access-Control-Request-Method request header needed", reqInfo, errors.GetAPIError(errors.ErrBadRequest))
+		return
+	}
+
+	headers := r.Header.Get(api.AccessControlRequestHeaders)
+
+	info, err := h.obj.GetBucketCORS(r.Context(), bktInfo)
+	if err != nil {
+		h.logAndSendError(w, "could not get cors", reqInfo, err)
+		return
+	}
+	cors := &CORSConfiguration{}
+	if err = xml.Unmarshal(info, cors); err != nil {
+		h.logAndSendError(w, "could not parse cors configuration", reqInfo, err)
+		return
+	}
+
+	for _, rule := range cors.CORSRules {
+		for _, o := range rule.AllowedOrigins {
+			if o == origin || o == wildcard {
+				for _, m := range rule.AllowedMethods {
+					if m == method {
+						if !checkSubslice(rule.AllowedHeaders, strings.Split(headers, ", ")) {
+							continue
+						}
+						w.Header().Set(api.AccessControlAllowOrigin, o)
+						w.Header().Set(api.AccessControlAllowMethods, strings.Join(rule.AllowedMethods, ", "))
+						if headers != "" {
+							w.Header().Set(api.AccessControlAllowHeaders, strings.Join(rule.AllowedHeaders, ", "))
+						}
+						if rule.ExposeHeaders != nil {
+							w.Header().Set(api.AccessControlExposeHeaders, strings.Join(rule.ExposeHeaders, ", "))
+						}
+						if rule.MaxAgeSeconds != 0 {
+							w.Header().Set(api.AccessControlMaxAge, strconv.Itoa(rule.MaxAgeSeconds))
+						} else {
+							w.Header().Set(api.AccessControlMaxAge, h.cfg.DefaultMaxAge)
+						}
+						if o != wildcard {
+							w.Header().Set(api.AccessControlAllowCredentials, "true")
+						}
+						api.WriteSuccessResponseHeadersOnly(w)
+						return
+					}
+				}
+			}
+		}
+	}
+	h.logAndSendError(w, "Forbidden", reqInfo, errors.GetAPIError(errors.ErrAccessDenied))
```

Comment:
If I understand correctly cors we can should not forbid the request and check allowed methods. It seems we can just return `Access-Control-Allowed-Methods` and other headers. Browsers must handle this correctly. So we can drop the `sliceContains` and `checkSubslice` functions

Reply:
I'm not sure I understeand correctly, but here it says that wee need to check method in preflight.
https://docs.aws.amazon.com/AmazonS3/latest/userguide/cors.html

Also, when I tried to make preflight request for a method which is not allowed in the first mathed rule, I received: 
```
<?xml version="1.0" encoding="UTF-8"?>
<Error>
    <Code>AccessForbidden</Code>
    <Message>CORSResponse: This CORS request is not allowed. This is usually because the evalution of Origin, request method / Access-Control-Request-Method or Access-Control-Request-Headers are not whitelisted by the resource's CORS spec.</Message>
    <Method>POST</Method>
    <ResourceType>BUCKET</ResourceType>
    <RequestId>W3ZGMMHZ9FNHX78H</RequestId>
    <HostId>wCqJjotZlC+gInT77Ewoog4MvI0rH3U/RKhc4F720ob3pC6zHUCuteSlug/TFH2RD74iP5vuMvg=</HostId>
</Error>
```

Reply:
It looks like I was wrong